### PR TITLE
Refactor sparse tensor code

### PR DESCRIPTION
**Pull request by reminisce**
This PR refactors the sparse tensor code based upon the comments here. https://github.com/dmlc/mxnet/pull/5800

1. In the imperative mode, returns stypes of output ndarrays together with their handles to the front end to avoid the c-api for the stypes of those handles when constructing python NDArrays. 

2. Saves stype attribute in the python NDArray class to avoid c-api calls to the backend for retrieving the stype from an NDArray handle.

3. Created a folder named `ndarray` under `src/python/mxnet` as a subpackage for `ndarray.py`, `sparse_ndarray.py`, `ndarray_utils.py`, `op.py`, and `_internal.py`. The folder acts as the namespace `mxnet.ndarray` for all ndarray related functions.

4. Moved the parallel cpu implementation of element-wise sum of a vector of row-sparse matrices to `src/ndarray/ndarray_function.cc`.

5. Coding style fixes.

@piiswrong @eric-haibin-lin @madjam

### CODE REVIEW

File: python/mxnet/ndarray/ndarray.py [hunk 1]
```diff
@@ -748,8 +746,10 @@ def dtype(self):
         return _DTYPE_MX_TO_NP[mx_dtype.value]
 
     @property
-    def storage_type(self):
-        return _storage_type(self.handle)
+    def stype(self):
+        if self._stype is None:
+            self._stype = _storage_type(self.handle)
```
Comment (eric-haibin-lin) [inline python/mxnet/ndarray/ndarray.py]: We don't need `def _storage_type` anymore, you just just move the function body here

File: tests/python/unittest/test_sparse_ndarray.py [hunk 2]
```diff
@@ -96,17 +96,17 @@ def check_sparse_nd_prop_rsp():
...
-        rsp = mx.sparse_nd.row_sparse(values, indices, shape)
+        rsp = mx.nd.row_sparse(values, indices, shape)
         dns = mx.nd.zeros(shape)
         dns[1] = mx.nd.array(values[0])
         dns[3] = mx.nd.array(values[1])
-        #assert_almost_equal(rsp.asnumpy(), dns.asnumpy())
+        # assert_almost_equal(rsp.asnumpy(), dns.asnumpy())
         print('before', indices)
         print('mx', mx.nd.array(indices, dtype='int64')[1].asnumpy())
         indices_np = mx.nd.array(indices, dtype='int64').asnumpy()
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_sparse_ndarray.py]: the `print()`s can be removed

Comment (reminisce) [inline tests/python/unittest/test_sparse_ndarray.py]: Done.

File: python/mxnet/ndarray/ndarray.py (hunk 1 above)
Comment (reminisce) [inline python/mxnet/ndarray/ndarray.py]: This function is needed at _ndarray_cls as well. After discussion, we will keep it as is.

File: python/mxnet/ndarray/ndarray.py [hunk 3]
```diff
@@ -18,33 +18,37 @@
 
 import operator
 import numpy as np
-from .base import _LIB, numeric_types, OpHandle, c_str
-from .base import c_array, py_str, mx_real_t, _Null  # pylint: disable=unused-import
-from .base import mx_uint, NDArrayHandle, check_call
-from .base import ctypes2buffer
-from .context import Context
+import mxnet as mx
```
Comment (piiswrong) [inline python/mxnet/ndarray/ndarray.py]: don't do this.
Use relative import

Comment (reminisce) [inline python/mxnet/ndarray/ndarray.py]: Fixed.

File: python/mxnet/ndarray/op.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    ndarray_function.__name__ = func_name
+    ndarray_function.__doc__ = doc_str
+    ndarray_function.__module__ = 'mxnet.ndarray'
+    return ndarray_function
+
+
+# pylint: enable=too-many-locals, invalid-name
+def _init_ndarray_module(root_namespace):
+    """List and add all the ndarray functions to current module."""
+    print "calling _init_ndarray_module"
```
Comment (eric-haibin-lin) [inline python/mxnet/ndarray/op.py]: please remove this line

Comment (reminisce) [inline python/mxnet/ndarray/op.py]: Done.

File: python/mxnet/random.py [hunk 5]
```diff
@@ -5,13 +5,13 @@
...
 import ctypes
 from .base import _LIB, check_call
-from ._ndarray_internal import _sample_uniform as uniform
-from ._ndarray_internal import _sample_normal as normal
-from ._ndarray_internal import _sample_gamma as gamma
-from ._ndarray_internal import _sample_exponential as exponential
-from ._ndarray_internal import _sample_poisson as poisson
-from ._ndarray_internal import _sample_negbinomial as negative_binomial
-from ._ndarray_internal import _sample_gennegbinomial as generalized_negative_binomial
+from .ndarray._ndarray_internal import _sample_uniform as uniform
```
Comment (piiswrong) [inline python/mxnet/random.py]: rename _ndarray_internal.py to _internal.py

**Comment by piiswrong** [comment]
I'm starting to think storage_type might be better...

File: python/mxnet/random.py (hunk 5 above)
Comment (reminisce) [inline python/mxnet/random.py]: Done.

**Pull request merged**